cmd/podman/containers/cp.go

```diff
@@ -113,6 +116,55 @@ func doCopy(funcA func() error, funcB func() error) error {
 	return errorhandling.JoinErrors(copyErrors)
 }
 
+func copyContainerToContainer(fromContainer string, sourcePath string, toContainer string, destPath string) error {
+	if err := containerMustExist(fromContainer); err != nil {
+		return err
+	}
+
+	if err := containerMustExist(toContainer); err != nil {
+		return err
+	}
+
+	fromContainerInfo, err := registry.ContainerEngine().ContainerStat(registry.GetContext(), fromContainer, sourcePath)
+	if err != nil {
+		return errors.Wrapf(err, "%q could not be found on container %s", sourcePath, fromContainer)
+	}
+
+	toContainerInfo, err := registry.ContainerEngine().ContainerStat(registry.GetContext(), toContainer, destPath)
+	if err != nil {
+		return errors.Wrapf(err, "%q could not be found on container %s", destPath, toContainer)
```

Comment:
That should not be an error per-se.  If I do `cp 1:/foo.txt 2:/foo.txt` then "foo.txt" should be created at the destination.  Have a look at the copy functions above for the wiring.

Reply:
I am on it, I didn't understand why we are checking for this condition here: https://github.com/containers/podman/blob/master/cmd/podman/containers/cp.go#L266
If possible, could you please give an example

Reply:
If we do a `cp 1:/dir 2:/tmp/foo/` `foo` must exist. If we do a `cp 1:/dir 2:/tmp/foo` `foo` would created if it doesn't exist.

The condition enforces that rule.  The `podman-cp` man pages mention these rules.

Reply:
Thanks for adding those comments btw, they are very helpful! I will work on the missing cases soon

Reply:
Happy to help! Feel free to ping me here on IRC at any time.  I am reachable during regular CEST working hours.

Reply:
Sorry for the late reply, I didn't get time to work on this. 
I also haven't been able to get IRC to work for some reason. But anyway, I will ask here 
So I am confused about the case where a file exists on one container say ok.txt and then it is being copied like:
`podman container cp distracted_tesla:ok.txt pedantic_hertz:ok.txt` or `podman container cp distracted_tesla:ok.txt pedantic_hertz:abdcd` 
Error:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x166ae92]

goroutine 1 [running]:
```
How will I go about handling this? Can you point me in the right direction in the existing code?

And then, say a file exists in both containers then this error occurs:
```
[mehul@fedora bin]$ ./podman container cp distracted_tesla:ok.txt pedantic_hertz:/
[mehul@fedora bin]$ ./podman container cp distracted_tesla:ok.txt pedantic_hertz:ok.txt
Error: 1 error occurred:
        * error copying to container: copier: put: /ok.txt (/ok.txt): exists but is not a directory
```
Not completely sure how it is being handled for container to host and host to container copy but is the buildah package playing a role in this? Any hints?

Thanks!

Reply:
> Sorry for the late reply, I didn't get time to work on this.
> I also haven't been able to get IRC to work for some reason. But anyway, I will ask here
> So I am confused about the case where a file exists on one container say ok.txt and then it is being copied like:
> `podman container cp distracted_tesla:ok.txt pedantic_hertz:ok.txt` or `podman container cp distracted_tesla:ok.txt pedantic_hertz:abdcd`
> Error:
> 
> ```
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x166ae92]
> 
> goroutine 1 [running]:
> ```
> 
> How will I go about handling this? Can you point me in the right direction in the existing code?

Below `goroutine 1 [...]` should be a stack trace that points to the exact line of code of the nil derref.

> And then, say a file exists in both containers then this error occurs:
> 
> ```
> [mehul@fedora bin]$ ./podman container cp distracted_tesla:ok.txt pedantic_hertz:/
> [mehul@fedora bin]$ ./podman container cp distracted_tesla:ok.txt pedantic_hertz:ok.txt
> Error: 1 error occurred:
>         * error copying to container: copier: put: /ok.txt (/ok.txt): exists but is not a directory
> ```
> 
> Not completely sure how it is being handled for container to host and host to container copy but is the buildah package playing a role in this? Any hints?

container->host and host->container seems to allows that but it should fail. Looks like you found a bug :)

Reply:
I can tackle that bug. Thanks for finding that!

Reply:
Opened https://github.com/containers/podman/pull/10772

Reply:
Thank you for fixing that!

Reply:
@vrothberg a bit confused about git, how do I get the changes you made to the branch on my fork? 
would it be like checking out to main and then pulling the changes and checking out to my branch `c2c-copy` and then merging main to my branch? 
wouldn't that mean losing the changes I made?

Reply:
What I usually do is the following (`upstream` is the remote branch pointing to `github.com/containers/podman`):
```
$ git remote update
$ git pull --rebase upstream master
```

Note that while rebasing you may have to resolve merge conflicts.

Reply:
been on this for hours now, cant seem to figure this. I was able to update the c2c-copy branch to the recent changes on podman but the changes you made yesterday don't reflect? even git log shows your merged PR

I think I might have to just re-add those changes and continue

Reply:
Nevermind, I figured it out. Didn't have to do anything manually

Reply:
@vrothberg so I am not sure how we will go about renaming files for container-container copy in a case like:
`podman container cp distracted_tesla:ok.txt pedantic_hertz:ok.txt`
In host->container or container->host I can see that the buildah copier package is handling that for us but how do we do this for this case?

I was able to fix the segmentation fault 
```
[mehul@fedora bin]$ ./podman cp c1:file.txt c2:/skbxlswx
[mehul@fedora bin]$ ./podman exec c2 /bin/sh -c "ls"
bin
dev
etc
file.txt
home
proc
root
run
sys
tmp
usr
var
```